File: app/src/main/java/com/stevenfrew/beatprompter/LineMeasurements.java
package com.stevenfrew.beatprompter;

import java.util.ArrayList;

public class LineMeasurements
{
    int mLineWidth;
    int mLineHeight;
    long[] mPixelsToTimes;
    int[] mGraphicHeights;
    int[] mJumpScrollIntervals=new int[101];
    int mLines;
    int mHighlightColour;

    LineMeasurements(int lines,int lineWidth,int lineHeight,ArrayList<Integer> graphicHeights,int highlightColour,LineEvent lineEvent,Line nextLine,long yStartScrollTime,ScrollingMode scrollMode)
    {
        mLines=lines;
        mLineHeight=lineHeight;
        mLineWidth=lineWidth;
        mGraphicHeights=new int[graphicHeights.size()];
        for(int f=0;f<mGraphicHeights.length;++f)
            mGraphicHeights[f]=graphicHeights.get(f);
        mHighlightColour=highlightColour;

        for(int f=0;f<101;++f)
        {
            double percentage=((double)f/100.0);
            mJumpScrollIntervals[f]=Math.min((int)((double)lineHeight*Utils.mSineLookup[(int)(90.0*percentage)]),lineHeight);
        }

        mPixelsToTimes=new long[Math.max(1, lineHeight)];
        long lineStartTime=lineEvent.mEventTime;
        long lineEndTime=lineEvent.mEventTime+((scrollMode==ScrollingMode.Smooth || nextLine!=null)?lineEvent.mDuration:0);
        long timeDiff=lineEndTime-yStartScrollTime;
        mPixelsToTimes[0]=lineStartTime;
        for(int f=1;f<lineHeight;++f)
        {
            double linePercentage = ((double) f / (double) lineHeight);
            if(scrollMode==ScrollingMode.Beat) {
                int sineLookup = (int) (90.0 * linePercentage);
                long sineTimeDiff = (long) (timeDiff * Utils.mSineLookup[sineLookup]);
                mPixelsToTimes[f] = yStartScrollTime + sineTimeDiff;
            }
            else
            {
                long pixelTimeDiff=(long)(linePercentage*(double)timeDiff);
                mPixelsToTimes[f]=yStartScrollTime+pixelTimeDiff;
            }
        }
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/LineMeasurements.kt
package com.stevenfrew.beatprompter

import android.graphics.Rect
import com.stevenfrew.beatprompter.util.Utils

class LineMeasurements internal constructor(internal var mLines: Int, internal var mLineWidth: Int, internal var mLineHeight: Int, internal val mGraphicHeights: IntArray, lineTime:Long, lineDuration: Long, yStartScrollTime: Long, scrollMode: ScrollingMode) {
    internal var mPixelsToTimes: LongArray
    internal var mJumpScrollIntervals = IntArray(101)
    internal val mGraphicRectangles=mGraphicHeights.map { Rect(0,0,mLineWidth,it) }.toTypedArray()

    init {
        for (f in 0..100)
            mJumpScrollIntervals[f] = Math.min((mLineHeight.toDouble() * Utils.mSineLookup[(90.0 * (f.toDouble() / 100.0)).toInt()]).toInt(), mLineHeight)

        mPixelsToTimes = LongArray(Math.max(1, mLineHeight))
        val lineEndTime = lineTime + lineDuration
        val timeDiff = lineEndTime - yStartScrollTime
        mPixelsToTimes[0] = lineTime
        for (f in 1 until mLineHeight) {
            val linePercentage = f.toDouble() / mLineHeight.toDouble()
            val diff=
                if (scrollMode == ScrollingMode.Beat)
                    (timeDiff * Utils.mSineLookup[(90.0 * linePercentage).toInt()]).toLong()
                else
                    (linePercentage * timeDiff.toDouble()).toLong()
            mPixelsToTimes[f] = yStartScrollTime + diff
        }
    }

    fun findClosestEarliestPixel(time:Long):Int
    {
        return findClosestEarliestPixel(time,0,mPixelsToTimes.size-1,0)
    }
    private fun findClosestEarliestPixel(time:Long,left:Int,right:Int,bestIndex:Int):Int
    {
        if(left>right)
            return bestIndex
        val currentBestVal=mPixelsToTimes[bestIndex]
        val mid=(left+right)/2
        val midVal=mPixelsToTimes[mid]
        return if(midVal>time || time-midVal>time-currentBestVal)
            findClosestEarliestPixel(time,left,mid-1, bestIndex)
        else //if(midVal<time && time-midVal<time-currentBestVal)
            findClosestEarliestPixel(time,mid+1,right,mid)
    }
}
